### PR TITLE
Change imports for Node.js compatibility

### PR DESCRIPTION
**Pull request by Johennes**
Fixes: #4287

This is a variation of #4375 that uses `.ts` extensions and https://babeljs.io/docs/babel-preset-typescript#rewriteimportextensions.

## Checklist

-   [ ] Tests written for new code (and old code if feasible).
-   [ ] New or updated `public`/`exported` symbols have accurate [TSDoc](https://tsdoc.org/) documentation.
-   [x] Linter and other CI checks pass.
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md)).

### CODE REVIEW

File: babel.config.cjs [hunk 1]
```diff
@@ -14,7 +14,12 @@ module.exports = {
                 modules: process.env.NODE_ENV === "test" ? "commonjs" : false,
             },
         ],
-        "@babel/preset-typescript",
+        [
+            "@babel/preset-typescript",
+            {
+                rewriteImportExtensions: process.env.NODE_ENV !== "test",
```
Comment (richvdh) [inline babel.config.cjs:20]: might be good to have a comment here explaining why we set this at all, and why we *don't* set it for jest.

Comment (Johennes) [reply babel.config.cjs:22]: Good idea. Have added one.

File: babel.config.cjs [hunk 2]
```diff
@@ -14,7 +14,14 @@ module.exports = {
                 modules: process.env.NODE_ENV === "test" ? "commonjs" : false,
             },
         ],
-        "@babel/preset-typescript",
+        [
+            "@babel/preset-typescript",
+            {
+                // Node.js requires .js extensions on ESM imports. Our tests, however, still
+                // run in CommonJS (see above).
```
Comment (richvdh) [inline babel.config.cjs:21]:
```suggestion
                // Node.js requires .js extensions on ESM imports. However, `jest` uses the TS source (via babel)
                // and fails to resolve the `.js` names. To resolve this, we use the `.ts` names in the source, and
                // rewrite the extensions for `.js` during transpilation, *except* when we are targetting jest.
```

Comment (Johennes) [reply babel.config.cjs]: Sorry yes, I was being sloppy there. This is much better.

**Comment by Johennes** [comment]
For resolving conflicts from here on, it should only be a matter of running the script from https://github.com/matrix-org/matrix-js-sdk/pull/4375 again (after replacing `js` with `ts`). Happy to do it if you give me a ping whenever you'd like to land this.

**Comment by richvdh** [comment]
@matrix-org/element-call-reviewers : would really love a review on this please...

**Review by AndrewFerr — APPROVED** [review]

**Comment by florianduros** [comment]
This is breaking the CI of the react sdk https://github.com/matrix-org/matrix-react-sdk/actions/runs/10627716730/job/29461292604?pr=12938

**Comment by richvdh** [comment]
Sigh. Annoying that it didn't get picked up by the CI on this PR. I guess you need to set `allowImportingTsExtensions` in the react-sdk too, if you're going to continue to use the TS source rather than the transpiled js.

**Comment by florianduros** [comment]
Sadly it's not enought https://github.com/matrix-org/matrix-react-sdk/pull/12939 I'm still investigating

**Comment by Johennes** [comment]
> This is breaking the CI of the react sdk https://github.com/matrix-org/matrix-react-sdk/actions/runs/10627716730/job/29461292604?pr=12938

Ugh, sorry and thanks for fixing this @florianduros.

**Pull request merged**